docs/formal-review.md

```diff
@@ -0,0 +1,146 @@
+# Formal review checklist and considerations
+
+> "The cost of fixing a defect is proportional to the time between defect introduction and discovery" - main axiom of softwaer engineering.
+
+## Purpose of the document and intented audience
+
+The purpose of this document is to list common defects and their sources to prevent their introduction into a code base.
+
+It is advisory for Columnstore developers to consult this document while doing personal erview of the code to be published. We also should extend this document as new sources of defects come to light.
+
+
+# Performance related issues
+
+## Assignment: x = y;
+
+If x is a std::string and y is a pointer, y is treated as an ASCIIZ string. There is implicit strlen call in there.
+
+
+# Side effects
+
```

Comment:
If a function introduces unexpected side-effects they must be explicitly described in the comments so that programmers in future are notified about side-effects.